File: CS/Reporting.Generation.Console/ConsoleAIReportGenerationHost.cs
using DevExpress.AIIntegration.Reporting;

namespace Reporting.Generation.Console {
    // Console host for interactive Prompt-to-Report generation.
    public class ConsoleAIReportGenerationHost : IAIReportGenerationHost {
        // Tracks where dynamic status lines are printed.
        private int currentCursorTop;
        // Indicates whether status output is printed for the first time.
        private bool isFirstStatusLine = true;
        // Stores the latest status value to detect status changes.
        private string lastStatus = string.Empty;
        private const int MenuInstructionsOffset = 2;
        // Request clarification from the user (choice list or free text).
        public Task<PromptClarificationAnswer> ClarifyPromptAsync(PromptClarificationQuestion request) {
            ClearStatusLines();
            DisplayQuestion(request.Text);
            if (request.Choices != null && request.Choices.Count > 0) {
                return HandleChoiceQuestion(request.Choices);
            } else {
                return HandleTextInputQuestion();
            }
        }
        // Clear previously rendered status lines before showing a question.
        void ClearStatusLines() {
            if (!isFirstStatusLine) {
                try {
                    ClearLine(currentCursorTop - 1);
                    ClearLine(currentCursorTop);
                    System.Console.SetCursorPosition(0, currentCursorTop - 1);
                } catch {
                    System.Console.WriteLine();
                }
            }
        }
        // Clear a single console line by overwriting it with spaces.
        void ClearLine(int lineNumber) {
            System.Console.SetCursorPosition(0, lineNumber);
            System.Console.Write(new string(' ', System.Console.BufferWidth - 1));
        }
        // Display a clarification question in a highlighted color.
        void DisplayQuestion(string questionText) {
            System.Console.ForegroundColor = ConsoleColor.Cyan;
            System.Console.WriteLine();
            System.Console.WriteLine(questionText);
            System.Console.ResetColor();
            System.Console.WriteLine();
        }
        // Handle clarification questions with predefined options.
        Task<PromptClarificationAnswer> HandleChoiceQuestion(IReadOnlyList<string> choices) {
            var selectedIndex = ShowInteractiveMenu(choices);
            isFirstStatusLine = true;
            if (selectedIndex == -1) {
                return Task.FromResult(PromptClarificationAnswer.Canceled());
            }
            return Task.FromResult(PromptClarificationAnswer.FromValue(choices[selectedIndex]));
        }
        // Handle clarification questions that require a text response.
        Task<PromptClarificationAnswer> HandleTextInputQuestion() {
            System.Console.ForegroundColor = ConsoleColor.DarkGray;
            System.Console.WriteLine("(Press Enter without text to cancel)");
            System.Console.ResetColor();
            System.Console.Write("Your answer: ");
            var answer = System.Console.ReadLine();
            System.Console.WriteLine();
            isFirstStatusLine = true;
            if (string.IsNullOrWhiteSpace(answer)) {
                return Task.FromResult(PromptClarificationAnswer.Canceled());
            }
            return Task.FromResult(PromptClarificationAnswer.FromValue(answer));
        }
        // Render an interactive menu and return the selected option index.
        int ShowInteractiveMenu(IReadOnlyList<string> choices) {
            int selectedIndex = 0;
            var startTop = System.Console.CursorTop;
            System.Console.WriteLine("Use ↑↓ arrows to navigate, Enter to select, Esc to cancel");
            System.Console.WriteLine();
            while (true) {
                RenderMenuItems(choices, selectedIndex, startTop);
                var action = ProcessKeyInput(ref selectedIndex, choices.Count);
                if (action.HasValue) {
                    System.Console.SetCursorPosition(0, startTop + MenuInstructionsOffset + choices.Count);
                    System.Console.WriteLine();
                    return action.Value;
                }
            }
        }
        // Draw all menu items for the current selection state.
        void RenderMenuItems(IReadOnlyList<string> choices, int selectedIndex, int startTop) {
            for (int i = 0; i < choices.Count; i++) {
                System.Console.SetCursorPosition(0, startTop + MenuInstructionsOffset + i);
                RenderMenuItem(choices[i], i == selectedIndex);
            }
        }
        // Draw one menu item with highlighted styling for the selected row.
        void RenderMenuItem(string text, bool isSelected) {
            if (isSelected) {
                System.Console.BackgroundColor = ConsoleColor.Gray;
                System.Console.ForegroundColor = ConsoleColor.Black;
                System.Console.Write($"> {text}");
                System.Console.ResetColor();
            } else {
                System.Console.Write($"  {text}");
            }
            var padding = Math.Max(0, System.Console.BufferWidth - text.Length - 3);
            System.Console.Write(new string(' ', padding));
        }
        // Process keyboard input and update the selected item.
        int? ProcessKeyInput(ref int selectedIndex, int itemCount) {
            var key = System.Console.ReadKey(true);
            switch (key.Key) {
                case ConsoleKey.UpArrow:
                    selectedIndex = selectedIndex > 0 ? selectedIndex - 1 : itemCount - 1;
                    return null;
                case ConsoleKey.DownArrow:
                    selectedIndex = selectedIndex < itemCount - 1 ? selectedIndex + 1 : 0;
                    return null;
                case ConsoleKey.Enter:
                    return selectedIndex;
                case ConsoleKey.Escape:
                    return -1;
                default:
                    return null;
            }
        }
        // Receive progress updates from the report generation workflow.
        public void NotifyAsync(string status, string reasoning) {
            bool statusChanged = lastStatus != status;
            lastStatus = status;
            if (isFirstStatusLine) {
                DisplayInitialStatus(status, reasoning);
            } else {
                UpdateStatus(status, reasoning, statusChanged);
            }
        }
        // Print initial status and reasoning lines.
        void DisplayInitialStatus(string status, string reasoning) {
            System.Console.WriteLine($"Status: {status}");
            if (!string.IsNullOrEmpty(reasoning)) {
                System.Console.Write($"Reasoning: {reasoning}");
            }
            currentCursorTop = System.Console.CursorTop;
            isFirstStatusLine = false;
        }
        // Update status output in-place when possible.
        void UpdateStatus(string status, string reasoning, bool statusChanged) {
            try {
                if (statusChanged) {
                    UpdateBothLines(status, reasoning);
                } else {
                    UpdateReasoningOnly(reasoning);
                }
            } catch {
                FallbackStatusDisplay(status, reasoning);
            }
        }
        // Re-render both status and reasoning lines.
        void UpdateBothLines(string status, string reasoning) {
            ClearLine(currentCursorTop - 1);
            System.Console.SetCursorPosition(0, currentCursorTop - 1);
            System.Console.WriteLine($"Status: {status}");

            ClearLine(currentCursorTop);
            System.Console.SetCursorPosition(0, currentCursorTop);
            if (!string.IsNullOrEmpty(reasoning)) {
                System.Console.Write($"Reasoning: {reasoning}");
            }
        }
        // Re-render only the reasoning line when status text has not changed.
        void UpdateReasoningOnly(string reasoning) {
            ClearLine(currentCursorTop);
            System.Console.SetCursorPosition(0, currentCursorTop);
            if (!string.IsNullOrEmpty(reasoning)) {
                System.Console.Write($"Reasoning: {reasoning}");
            }
        }
        // Fall back to plain output when cursor-based updates fail.
        void FallbackStatusDisplay(string status, string reasoning) {
            System.Console.WriteLine();
            System.Console.WriteLine($"Status: {status}");
            if (!string.IsNullOrEmpty(reasoning)) {
                System.Console.Write($"Reasoning: {reasoning}");
            }
            currentCursorTop = System.Console.CursorTop;
        }
    }
}

File: VB/Reporting.Generation.Console/ConsoleAIReportGenerationHost.vb
Imports DevExpress.AIIntegration.Reporting

Namespace Reporting.Generation.Console

    ' Console host for interactive Prompt-to-Report generation.
    Public Class ConsoleAIReportGenerationHost
        Implements IAIReportGenerationHost

        ' Tracks where dynamic status lines are printed.
        Private currentCursorTop As Integer

        ' Indicates whether status output is printed for the first time.
        Private isFirstStatusLine As Boolean = True

        ' Stores the latest status value to detect status changes.
        Private lastStatus As String = String.Empty

        Private Const MenuInstructionsOffset As Integer = 2

        ' Request clarification from the user (choice list or free text).
        Public Function ClarifyPromptAsync(ByVal request As PromptClarificationQuestion) As Task(Of PromptClarificationAnswer) Implements IAIReportGenerationHost.ClarifyPromptAsync
            ClearStatusLines()
            Me.DisplayQuestion(request.Text)
            If request.Choices IsNot Nothing AndAlso request.Choices.Count > 0 Then
                Return HandleChoiceQuestion(request.Choices)
            Else
                Return HandleTextInputQuestion()
            End If
        End Function

        ' Clear previously rendered status lines before showing a question.
        Private Sub ClearStatusLines()
            If Not isFirstStatusLine Then
                Try
                    ClearLine(currentCursorTop - 1)
                    Me.ClearLine(currentCursorTop)
                    System.Console.SetCursorPosition(0, currentCursorTop - 1)
                Catch
                    System.Console.WriteLine()
                End Try
            End If
        End Sub

        ' Clear a single console line by overwriting it with spaces.
        Private Sub ClearLine(ByVal lineNumber As Integer)
            System.Console.SetCursorPosition(0, lineNumber)
            System.Console.Write(New String(" "c, System.Console.BufferWidth - 1))
        End Sub

        ' Display a clarification question in a highlighted color.
        Private Sub DisplayQuestion(ByVal questionText As String)
            System.Console.ForegroundColor = ConsoleColor.Cyan
            System.Console.WriteLine()
            System.Console.WriteLine(questionText)
            System.Console.ResetColor()
            System.Console.WriteLine()
        End Sub

        ' Handle clarification questions with predefined options.
        Private Function HandleChoiceQuestion(ByVal choices As IReadOnlyList(Of String)) As Task(Of PromptClarificationAnswer)
            Dim selectedIndex = Me.ShowInteractiveMenu(choices)
            isFirstStatusLine = True
            If selectedIndex = -1 Then
                Return Task.FromResult(PromptClarificationAnswer.Canceled())
            End If

            Return Task.FromResult(PromptClarificationAnswer.FromValue(choices(selectedIndex)))
        End Function

        ' Handle clarification questions that require a text response.
        Private Function HandleTextInputQuestion() As Task(Of PromptClarificationAnswer)
            System.Console.ForegroundColor = ConsoleColor.DarkGray
            System.Console.WriteLine("(Press Enter without text to cancel)")
            System.Console.ResetColor()
            System.Console.Write("Your answer: ")
            Dim answer = System.Console.ReadLine()
            System.Console.WriteLine()
            isFirstStatusLine = True
            If String.IsNullOrWhiteSpace(answer) Then
                Return Task.FromResult(PromptClarificationAnswer.Canceled())
            End If

            Return Task.FromResult(PromptClarificationAnswer.FromValue(answer))
        End Function

        ' Render an interactive menu and return the selected option index.
        Private Function ShowInteractiveMenu(ByVal choices As IReadOnlyList(Of String)) As Integer
            Dim selectedIndex As Integer = 0
            Dim startTop = System.Console.CursorTop
            System.Console.WriteLine("Use ↑↓ arrows to navigate, Enter to select, Esc to cancel")
            System.Console.WriteLine()
            While True
                RenderMenuItems(choices, selectedIndex, startTop)
                Dim action = ProcessKeyInput(selectedIndex, choices.Count)
                If action.HasValue Then
                    System.Console.SetCursorPosition(0, startTop + MenuInstructionsOffset + choices.Count)
                    System.Console.WriteLine()
                    Return action.Value
                End If
            End While
        End Function

        ' Draw all menu items for the current selection state.
        Private Sub RenderMenuItems(ByVal choices As IReadOnlyList(Of String), ByVal selectedIndex As Integer, ByVal startTop As Integer)
            For i As Integer = 0 To choices.Count - 1
                System.Console.SetCursorPosition(0, startTop + MenuInstructionsOffset + i)
                RenderMenuItem(choices(i), i = selectedIndex)
            Next
        End Sub

        ' Draw one menu item with highlighted styling for the selected row.
        Private Sub RenderMenuItem(ByVal text As String, ByVal isSelected As Boolean)
            If isSelected Then
                System.Console.BackgroundColor = ConsoleColor.Gray
                System.Console.ForegroundColor = ConsoleColor.Black
                System.Console.Write($"> {text}")
                System.Console.ResetColor()
            Else
                System.Console.Write($"  {text}")
            End If

            Dim padding = Math.Max(0, System.Console.BufferWidth - text.Length - 3)
            System.Console.Write(New String(" "c, padding))
        End Sub

        ' Process keyboard input and update the selected item.
        Private Function ProcessKeyInput(ByRef selectedIndex As Integer, ByVal itemCount As Integer) As Integer?
            Dim key = System.Console.ReadKey(True)
            Select Case key.Key
                Case ConsoleKey.UpArrow
                    selectedIndex = If(selectedIndex > 0, selectedIndex - 1, itemCount - 1)
                    Return Nothing
                Case ConsoleKey.DownArrow
                    selectedIndex = If(selectedIndex < itemCount - 1, selectedIndex + 1, 0)
                    Return Nothing
                Case ConsoleKey.Enter
                    Return selectedIndex
                Case ConsoleKey.Escape
                    Return -1
                Case Else
                    Return Nothing
            End Select
        End Function

        ' Receive progress updates from the report generation workflow.
        Public Sub NotifyAsync(ByVal status As String, ByVal reasoning As String) Implements IAIReportGenerationHost.NotifyAsync
            Dim statusChanged As Boolean = Not Equals(lastStatus, status)
            lastStatus = status
            If isFirstStatusLine Then
                Me.DisplayInitialStatus(status, reasoning)
            Else
                Me.UpdateStatus(status, reasoning, statusChanged)
            End If
        End Sub

        ' Print initial status and reasoning lines.
        Private Sub DisplayInitialStatus(ByVal status As String, ByVal reasoning As String)
            System.Console.WriteLine($"Status: {status}")
            If Not String.IsNullOrEmpty(reasoning) Then
                System.Console.Write($"Reasoning: {reasoning}")
            End If

            currentCursorTop = System.Console.CursorTop
            isFirstStatusLine = False
        End Sub

        ' Update status output in-place when possible.
        Private Sub UpdateStatus(ByVal status As String, ByVal reasoning As String, ByVal statusChanged As Boolean)
            Try
                If statusChanged Then
                    Me.UpdateBothLines(status, reasoning)
                Else
                    Me.UpdateReasoningOnly(reasoning)
                End If
            Catch
                Me.FallbackStatusDisplay(status, reasoning)
            End Try
        End Sub

        ' Re-render both status and reasoning lines.
        Private Sub UpdateBothLines(ByVal status As String, ByVal reasoning As String)
            ClearLine(currentCursorTop - 1)
            System.Console.SetCursorPosition(0, currentCursorTop - 1)
            System.Console.WriteLine($"Status: {status}")
            Me.ClearLine(currentCursorTop)
            System.Console.SetCursorPosition(0, currentCursorTop)
            If Not String.IsNullOrEmpty(reasoning) Then
                System.Console.Write($"Reasoning: {reasoning}")
            End If
        End Sub

        ' Re-render only the reasoning line when status text has not changed.
        Private Sub UpdateReasoningOnly(ByVal reasoning As String)
            Me.ClearLine(currentCursorTop)
            System.Console.SetCursorPosition(0, currentCursorTop)
            If Not String.IsNullOrEmpty(reasoning) Then
                System.Console.Write($"Reasoning: {reasoning}")
            End If
        End Sub

        ' Fall back to plain output when cursor-based updates fail.
        Private Sub FallbackStatusDisplay(ByVal status As String, ByVal reasoning As String)
            System.Console.WriteLine()
            System.Console.WriteLine($"Status: {status}")
            If Not String.IsNullOrEmpty(reasoning) Then
                System.Console.Write($"Reasoning: {reasoning}")
            End If

            currentCursorTop = System.Console.CursorTop
        End Sub
    End Class
End Namespace
